src/lib/arch/riscv/Instruction_execute.cc

```diff
@@ -827,6 +841,642 @@ void Instruction::execute() {
+      const double rs2 = operands[1].get<double>();
+
+      results[0] = RegisterValue(std::copysign(rs1, rs2), 8);
+      break;
+    }
+    case Opcode::RISCV_FSGNJ_S: {  // FSGNJ.S rd,rs1,rs2
+      const float rs1 = operands[0].get<float>();
+      const float rs2 = operands[1].get<float>();
+
+      results[0] = RegisterValue(NanBoxFloat(std::copysign(rs1, rs2)), 8);
+      break;
+    }
+    case Opcode::RISCV_FSGNJN_D: {  // FSGNJN.D rd,rs1,rs2
+      const double rs1 = operands[0].get<double>();
+      const double rs2 = operands[1].get<double>();
+
+      results[0] = RegisterValue(std::copysign(rs1, -rs2), 8);
+      break;
+    }
+
+    case Opcode::RISCV_FSGNJN_S: {  // FSGNJN.S rd,rs1,rs2
+      const float rs1 = operands[0].get<float>();
+      const float rs2 = operands[1].get<float>();
+
+      results[0] = RegisterValue(std::copysign(rs1, -rs2), 8);
+      break;
+    }
+    case Opcode::RISCV_FSGNJX_D: {  // FSGNJX.D rd,rs1,rs2
+      const double rs1 = operands[0].get<double>();
+      const double rs2 = operands[1].get<double>();
+
+      const double xorSign = pow(-1, std::signbit(rs1) ^ std::signbit(rs2));
+
+      results[0] = RegisterValue(std::copysign(rs1, xorSign), 8);
+      break;
+    }
+    case Opcode::RISCV_FSGNJX_S: {  // FSGNJX.S rd,rs1,rs2
+      const float rs1 = operands[0].get<float>();
+      const float rs2 = operands[1].get<float>();
+
+      const float xorSign = pow(-1, std::signbit(rs1) ^ std::signbit(rs2));
+
+      results[0] = RegisterValue(NanBoxFloat(std::copysign(rs1, xorSign)), 8);
+      break;
+    }
+
+    case Opcode::RISCV_FMV_D_X: {  // FMV.D.X rd,rs1
+      const double rs1 = operands[0].get<double>();
+
+      results[0] = RegisterValue(rs1, 8);
+      break;
+    }
+    case Opcode::RISCV_FMV_X_D: {  // FMV.X.D rd,rs1
+      const double rs1 = operands[0].get<double>();
+
+      results[0] = RegisterValue(rs1, 8);
+      break;
+    }
+    case Opcode::RISCV_FMV_W_X: {  // FMV.W.X rd,rs1
+      const float rs1 = operands[0].get<float>();
+
+      results[0] = RegisterValue(NanBoxFloat(rs1), 8);
+      break;
+    }
+    case Opcode::RISCV_FMV_X_W: {  // FMV.X.W rd,rs1
+      const uint64_t rs1 = operands[0].get<uint64_t>();
+
+      results[0] = RegisterValue(signExtendW(rs1), 8);
+      break;
+    }
+
+      // TODO FLT.S and FLE.S perform what the IEEE 754-2008 standard refers to
+      // as signaling comparisons: that is, they set the invalid operation
+      // exception flag if either input is NaN. FEQ.S performs a quiet
+      // comparison: it only sets the invalid operation exception flag if either
+      // input is a signaling NaN. For all three instructions, the result is 0
+      // if either operand is NaN.
+    case Opcode::RISCV_FEQ_D: {  // FEQ.D rd,rs1,rs2
+      // TODO FEQ.S performs a quiet
+      // comparison: it only sets the invalid operation exception flag if
+      // either input is a signaling NaN. Qemu doesn't seem to set CSR flags
+      // with sNANs
```

Comment:
Is there something blocking this TODO?

Reply:
The control and status register implementation is not complete/non-existent and so does not support updating CSR values. The comment also adds that Qemu doesn't perform this behaviour and so I am not entirely sure what is the correct behaviour

Reply:
Updated comment